Ignore logger methods to avoid graph breaks

SNahir opened the pull request:
Fixes #132635

Calls to logging.logger cause a graph break, this PR allows the user to avoid these graph breaks (for specific methods) by setting DISABLE_LOGS_WHILE_COMPILING to 1.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

SNahir committed:
Ignore logger methods to avoid graph breaks

williamwen42 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -1460,7 +1460,14 @@ def call_method(
+            if name == "isEnabledFor":
+                return variables.ConstantVariable.create(False)
+            if name in ['info', 'debug', 'error', 'warning', 'critical', 'exception']:
+                return
```
We should be returning `ConstantVariable.create(None)` here (and above as well).

SNahir replied:
done

williamwen42 replied:
I meant that the empty return statements `return` should be replaced with `return ConstantVariable.create(None)` (there is another instance of this above this highlighted code that should also be changed. `variables.ConstantVariable.create(False)` should remain unchanged (although this is technically incorrect behavior).

williamwen42 reviewed (commented):
Can we combine functionality with `torch._dynamo.config.reorderable_logging_functions`?

SNahir committed:
fix - replace False with None

SNahir commented:
> Can we combine functionality with `torch._dynamo.config.reorderable_logging_functions`?

Hi @williamwen42 , I looked at https://github.com/pytorch/pytorch/pull/116106 
It looks like reorderable_logging_functions  is specifically for functions and not methods (it is mentioned in the pr but also is_reorderable_logging_function returns False for methods, call_method is not implemented in DebuggingVariable). I think it is possible to combine LoggingLoggerVariable into DebuggingVariable, but it will require a lot of modifications for DebuggingVariable (e.g. implement a separate call_method, creating a separation between reordering/ ignoring logs). I think it's best to keep them separated as they serve different purposes and different classes, but I would like to hear your thoughts on this

SNahir commented:
> I don't think we need to combine `LoggingLoggerVariable` and `DebuggingVariable`, but we can still use the single config `reorderable_logging_functions`. I'm thinking we can add methods at the class level (i.e. `logging.Logger.info`) and `LoggingLoggerVariable` can check if the method's `__func__` is in `reorderable_logging_functions`.

Thank you for the clarification
I think using `reorderable_logging_functions` for this can be confusing for several reasons, one of which is that the purpose of this commit is to ignore calls to logger methods and not reorder them in such a way that they are still being called. 
Also, I tried adding logging.Logger.info to `reorderable_logging_functions` as you suggested just to see if I can use it in the way you mentioned and also to check if logging.Logger methods can be reordered the same way as logging functions. However the object being traced is logging.Logger and not a callable like logging.info, so the tracing of logging and logging.Logger is done differently (ie logger methods can't be reordered in the same way as logging functions).
So I think it might be better that logging and logging.Logger handled separately, including having separate configs. I can create a separate set for logger methods I want to ignore, but I'll still need to check for the name of the method in call_method for isEnabledFor

williamwen42 commented:
Ideally we would be able to reorder logging.Logger methods rather than skip them, but I'm ok if we skip them for now in order to work around graph breaks (and thus keep the configs separate).

My thoughts on getting reordering logging.Logger methods to work:

For using `reorderable_logging_functions`, I was thinking that `LoggingLoggerVariable` would additionally look into this set (`if getattr(logging.Logger, name) in reorderable_logging_functions`?) before deciding to skip/reorder.

Is there anything preventing us from adding reconstruction logic to enable reordering logging? The code would look something like
```python
def gen_fn():
    codegen(logger_var)
    codegen.append_output(codegen.create_load_attr(meth_name))
codegen.add_push_null(gen_fn)
codegen(arg)
codegen.create_call_function(1, False)
```

majian4work commented:
> > I don't think we need to combine `LoggingLoggerVariable` and `DebuggingVariable`, but we can still use the single config `reorderable_logging_functions`. I'm thinking we can add methods at the class level (i.e. `logging.Logger.info`) and `LoggingLoggerVariable` can check if the method's `__func__` is in `reorderable_logging_functions`.
> 
> Thank you for the clarification I think using `reorderable_logging_functions` for this can be confusing for several reasons, one of which is that the purpose of this commit is to ignore calls to logger methods and not reorder them in such a way that they are still being called. Also, I tried adding logging.Logger.info to `reorderable_logging_functions` as you suggested just to see if I can use it in the way you mentioned and also to check if logging.Logger methods can be reordered the same way as logging functions. However the object being traced is logging.Logger and not a callable like logging.info, so the tracing of logging and logging.Logger is done differently (ie logger methods can't be reordered in the same way as logging functions). So I think it might be better that logging and logging.Logger handled separately, including having separate configs. I can create a separate set for logger methods I want to ignore, but I'll still need to check for the name of the method in call_method for isEnabledFor

if you want to skip the logger in case users have put it in `reorderable_logging_functions` , then you should bypass `DebuggingVarialbe.is_redoderable_logging_function ` with your config.

SNahir commented:
> > > I don't think we need to combine `LoggingLoggerVariable` and `DebuggingVariable`, but we can still use the single config `reorderable_logging_functions`. I'm thinking we can add methods at the class level (i.e. `logging.Logger.info`) and `LoggingLoggerVariable` can check if the method's `__func__` is in `reorderable_logging_functions`.
> > 
> > 
> > Thank you for the clarification I think using `reorderable_logging_functions` for this can be confusing for several reasons, one of which is that the purpose of this commit is to ignore calls to logger methods and not reorder them in such a way that they are still being called. Also, I tried adding logging.Logger.info to `reorderable_logging_functions` as you suggested just to see if I can use it in the way you mentioned and also to check if logging.Logger methods can be reordered the same way as logging functions. However the object being traced is logging.Logger and not a callable like logging.info, so the tracing of logging and logging.Logger is done differently (ie logger methods can't be reordered in the same way as logging functions). So I think it might be better that logging and logging.Logger handled separately, including having separate configs. I can create a separate set for logger methods I want to ignore, but I'll still need to check for the name of the method in call_method for isEnabledFor
> 
> if you want to skip the logger in case users have put it in `reorderable_logging_functions` , then you should bypass `DebuggingVarialbe.is_redoderable_logging_function ` with your config.

I'm not sure why, in what scenario will DebuggingVarialbe.is_redoderable_logging_function returns true for logger? 
Whether the user adds logging.logger or a logger method (like logger.info) to reorderable_logging_functions, DebuggingVarialbe.is_redoderable_logging_function still returns false

majian4work commented:
> > > > I don't think we need to combine `LoggingLoggerVariable` and `DebuggingVariable`, but we can still use the single config `reorderable_logging_functions`. I'm thinking we can add methods at the class level (i.e. `logging.Logger.info`) and `LoggingLoggerVariable` can check if the method's `__func__` is in `reorderable_logging_functions`.
> > > 
> > > 
> > > Thank you for the clarification I think using `reorderable_logging_functions` for this can be confusing for several reasons, one of which is that the purpose of this commit is to ignore calls to logger methods and not reorder them in such a way that they are still being called. Also, I tried adding logging.Logger.info to `reorderable_logging_functions` as you suggested just to see if I can use it in the way you mentioned and also to check if logging.Logger methods can be reordered the same way as logging functions. However the object being traced is logging.Logger and not a callable like logging.info, so the tracing of logging and logging.Logger is done differently (ie logger methods can't be reordered in the same way as logging functions). So I think it might be better that logging and logging.Logger handled separately, including having separate configs. I can create a separate set for logger methods I want to ignore, but I'll still need to check for the name of the method in call_method for isEnabledFor
> > 
> > 
> > if you want to skip the logger in case users have put it in `reorderable_logging_functions` , then you should bypass `DebuggingVarialbe.is_redoderable_logging_function ` with your config.
> 
> I'm not sure why, in what scenario will DebuggingVarialbe.is_redoderable_logging_function returns true for logger? Whether the user adds logging.logger or a logger method (like logger.info) to reorderable_logging_functions, DebuggingVarialbe.is_redoderable_logging_function still returns false

sorry， I have no idea about this.

SNahir committed:
fix + using a set instead of environment variable

SNahir commented:
> > > > > I don't think we need to combine `LoggingLoggerVariable` and `DebuggingVariable`, but we can still use the single config `reorderable_logging_functions`. I'm thinking we can add methods at the class level (i.e. `logging.Logger.info`) and `LoggingLoggerVariable` can check if the method's `__func__` is in `reorderable_logging_functions`.
> > > > 
> > > > 
> > > > Thank you for the clarification I think using `reorderable_logging_functions` for this can be confusing for several reasons, one of which is that the purpose of this commit is to ignore calls to logger methods and not reorder them in such a way that they are still being called. Also, I tried adding logging.Logger.info to `reorderable_logging_functions` as you suggested just to see if I can use it in the way you mentioned and also to check if logging.Logger methods can be reordered the same way as logging functions. However the object being traced is logging.Logger and not a callable like logging.info, so the tracing of logging and logging.Logger is done differently (ie logger methods can't be reordered in the same way as logging functions). So I think it might be better that logging and logging.Logger handled separately, including having separate configs. I can create a separate set for logger methods I want to ignore, but I'll still need to check for the name of the method in call_method for isEnabledFor
> > > 
> > > 
> > > if you want to skip the logger in case users have put it in `reorderable_logging_functions` , then you should bypass `DebuggingVarialbe.is_redoderable_logging_function ` with your config.
> > 
> > 
> > I'm not sure why, in what scenario will DebuggingVarialbe.is_redoderable_logging_function returns true for logger? Whether the user adds logging.logger or a logger method (like logger.info) to reorderable_logging_functions, DebuggingVarialbe.is_redoderable_logging_function still returns false
> 
> seems it doesn't work for `logger.info`. because "logger" is not callable, and it then creates `LoggingLoggerVariable`. guess `DebuggingVariable.is_reorderable_logging_function` only works for functions, not include methods.

yes, this is also mentioned in my comment to @williamwen42 from 4/11 and in the PR for DebuggingVariable. Also, the object that is being traced when calling logger.info is logging.logger, so even "object in torch._dynamo.config.reorderable_logging_functions" returns false. So overall even in case the user adds logger.info to recordable set instead of logging.info, it has no effect on the tracing

SNahir commented:
> Ideally we would be able to reorder logging.Logger methods rather than skip them, but I'm ok if we skip them for now in order to work around graph breaks (and thus keep the configs separate).
> 
> My thoughts on getting reordering logging.Logger methods to work:
> 
> For using `reorderable_logging_functions`, I was thinking that `LoggingLoggerVariable` would additionally look into this set (`if getattr(logging.Logger, name) in reorderable_logging_functions`?) before deciding to skip/reorder.
> 
> Is there anything preventing us from adding reconstruction logic to enable reordering logging? The code would look something like
> 
> ```python
> def gen_fn():
>     codegen(logger_var)
>     codegen.append_output(codegen.create_load_attr(meth_name))
> codegen.add_push_null(gen_fn)
> codegen(arg)
> codegen.create_call_function(1, False)
> ```

Thanks! I fixed call_method according to your comment and also changed the config to use a set rather than an environment variable, so the user can add methods directly to the set

SNahir commented:
Hi @williamwen42 , did you have a chance to review my commit from last week? I fixed the code according to your comment

williamwen42 reviewed (commented):
Looks good, can you add a test? Probably to `test/dynamo/test_reorder_logs.py` (even though we haven't done reordering yet, this change is related).

SNahir committed:
add tests

williamwen42 commented on torch/_dynamo/config.py:
```diff
@@ -442,6 +442,10 @@ def default_debug_dir_root():
 
+# A set of methods that will be ignored while tracing,
+# to prevent graph breaks.
+ignore_logger_methods: Set[Callable] = set()
```
Add a comment detailing that functions that go here should be in the form `logging.Logger.*` (I see in your test that `logger_obj.*` doesn't do anything)

williamwen42 commented on test/dynamo/test_reorder_logs.py:
```diff
@@ -9,8 +9,69 @@
+        (logger.isEnabledFor, f_isEnabledFor),
+        (logging.Logger.isEnabledFor, f_isEnabledFor),
+    ])
+    def test_ignore_logger(self, ignore_method, fn):
```
Can we combine these tests into 1, since most of the code is the same?

SNahir committed:
fix comment_ tests

balaabhijit commented:
@SNahir could you please explain how to use this? I am trying to use torch.compile with transformers and I have graph breaks due to logging.

williamwen42 commented:
Tests are failing - please address them before I can merge. Looks like you need to run the linter (`lintrunner -a`).

SNahir committed:
fix tests failure

SNahir commented:
> @SNahir could you please explain how to use this? I am trying to use torch.compile with transformers and I have graph breaks due to logging.

@balaabhijit You can ignore logging methods (like logger.info / logging.Logger.info) by adding them to ignore_logger_methods in config.py:
using a decorator - @torch._dynamo.config.patch(ignore_logger_methods ={logger.info})
using a context manager - with torch._dynamo.config.patch(ignore_logger_methods={logger.info}):
just adding the method directly to the configuration- torch._dynamo.config.ignore_logger_methods.add(logger.info)

SNahir committed:
fix tests failure1

SNahir committed:
fix linter error

SNahir commented:
Hi @williamwen42 , can you re-run the tests? I think the test failures for the latest commit are unrelated to this PR:
https://github.com/pytorch/pytorch/actions/runs/12138038613/job/33847795528
https://github.com/pytorch/pytorch/actions/runs/12138038613/job/33847795528
While jobs that run the relevant tests are completed successfully: 
https://github.com/pytorch/pytorch/actions/runs/12138038652/job/33847113312

williamwen42 commented:
@pytorchbot merge
